neo/Consensus/ConsensusService.cs

```diff
@@ -101,35 +111,130 @@ private void CheckCommits()
             }
         }
 
-        private void CheckExpectedView(byte view_number)
+        private void CheckExpectedView(byte viewNumber)
         {
-            if (context.ViewNumber == view_number) return;
-            if (context.ExpectedView.Count(p => p == view_number) >= context.M)
-            {
-                InitializeConsensus(view_number);
-            }
+            if (context.ViewNumber == viewNumber) return;
+            if (context.ChangeViewPayloads.Count(p => p != null && p.GetDeserializedMessage<ChangeView>().NewViewNumber == viewNumber) >= context.M)
+                InitializeConsensus(viewNumber);
         }
 
         private void CheckPreparations()
         {
-            if (context.Preparations.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
+            if (context.PreparationPayloads.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
             {
                 ConsensusPayload payload = context.MakeCommit();
                 Log($"send commit");
                 context.State |= ConsensusState.CommitSent;
                 store.Put(ContextSerializationPrefix, new byte[0], context.ToArray());
                 localNode.Tell(new LocalNode.SendDirectly { Inventory = payload });
+                // Set timer, so we will resend the commit in case of a networking issue
+                ChangeTimer(TimeSpan.FromSeconds(Blockchain.SecondsPerBlock));
                 CheckCommits();
             }
         }
 
-        private void InitializeConsensus(byte view_number)
+        private static ConsensusPayload[] GetChangeViewPayloadsFromRecoveryMessage(IConsensusContext context, ConsensusPayload payload, RecoveryMessage message)
```

Comment:
Maybe we should move these methods into the recovery message themselves. I'm going to make that change locally, and see how it looks.

Reply:
You'll need the parent payload. I'm going to move them into a helper class.

Reply:
I guess they should stay where they are since they are setting all the members in `ConsensusService`.

Reply:
It is cleaner

Reply:
I'll commit this change in a few minutes.

Reply:
No, I already did it and didn't have any problems.

Reply:
I just left the payload passed into the method.

Reply:
Do you like 9926b74?

Reply:
actually i think it should not pass in the `ConsensusContext`. I'm cleaning that up now.

Reply:
This is fine now.

Reply:
Ok then.